### PR TITLE
fix issues about NVP flow

### PR DESCRIPTION
**Pull request by SyanneLiu**
In this PR, I modify several issues related to NVPFlow.

- There was a bug in NeuralNet. The non-linear activation was only applied to the last layer while the gradient of the non-linear activation was only computed on the intermediate layers. 
```
        for l in range(self._layers):
            W = var_param[str(l)]
            x = np.dot(x, W)
            log_det_J += np.log(np.abs(np.dot(derivative(x), W.T).sum(axis = 1)))
        return self._nonlinearity(x), log_det_J
```
The bug is fixed now.
- Added a feature to NeuralNet. Now the intermediate layers and the last layer can use different activation functions. 
- Fixed several bugs in NVPFlow, concluding the computation of `log_density` and `entropy`(removed).
- Added weight decay to the StochasticGradientOptimizer.

### CODE REVIEW

File: viabel/optimization.py [hunk 1]
```diff
@@ -82,6 +82,8 @@ def optimize(self, n_iters, objective, init_param):
                     object_val, object_grad = objective(variational_param)
                     descent_dir = self.descent_direction(object_grad)
                     variational_param -= self._learning_rate * descent_dir
+                    if len(variational_param.shape) == 2:
```
Comment (jhuggins) [inline viabel/optimization.py:85]: This looks very heuristic. What's the reason for adding it? Also, instead of `len(variational_param.shape)` you can use `variational_param.ndim`

**Review by jhuggins — CHANGES_REQUESTED** [review]
Just one question/comment about your modification to the optimizer

File: viabel/optimization.py (hunk 1 above)
Comment (SyanneLiu) [reply viabel/optimization.py:85]: When training neural networks, people usually use weight decay (L2 regularization). I'm not sure if that is the case in variational inference, but I ran into numerical instability when fitting NVPFlow. That's the reason to add it. 

If it's fair to add weight decay to optimizers, I can add it as a parameter for all optimizers.

Comment (jhuggins) [reply viabel/optimization.py:85]: Yeah, why don't you add a weight decay option to the optimizer (as an argument to `__init__`).

Comment (SyanneLiu) [reply viabel/optimization.py]: Added.

**Review by jhuggins — APPROVED** [review]
